fixes #1215 replacing share menu by dialog

gerardolgvr opened the pull request:
**Description (required)**
replacing share menu by dialog

Fixes #1215 

**What changes did you make and why?**

- New ShareDialog created 
- Updating menus according to the new dialog (feeditems_options.xml, mediapalyer.xml)
- Refactor where the share links where accesed (FeedItemMenuHandler and MediaPlayerActivity)

Tested with freeDebug on Redmi Note 7 with API level 28 and Nexus 5X API level 28 (emulator).

**Screenshots**
![WhatsApp Image 2020-06-23 at 5 24 09 PM](https://user-images.githubusercontent.com/20119298/85471706-71e7b380-b576-11ea-92e4-2f16b65b32ca.jpeg)
![WhatsApp Image 2020-06-23 at 5 24 09 PM (1)](https://user-images.githubusercontent.com/20119298/85471724-79a75800-b576-11ea-81ce-e40754415557.jpeg)


I read some comments about the strings that contains "URL". Should I update and change "URL" for "Link"? If there's something wrong feel free to let me know. I will do the necesary changes

**Notes**
- The dialog is showed when you click the option "Share..." at menu and the info about the episode just as before. The difference is now it shows the dialog instead the submenu.

gerardolgvr commented:
Hello @ByteHamster , is there any section related to tests for dialogs? 😅

ByteHamster commented:
I don't think we currently have a test for dialogs. If you find a good test, feel free to add one :)

#1215 contains a mock-up of a different dialog design. The advantage of that dialog is that the text that is displayed in one line can be a lot shorter. To completely fix that issue, the dialog should look more like the mock-ups there. The long text is a problem on small screens, as well as languages where the translation requires more characters (German, for example).

gerardolgvr commented:
Ohh you're right! At the beginning, I was a little confused if it should look like the mockup. 

Yup, I can update it and try to find the way to test the dialog. 

At #1215  I saw 2 dialogs, one corresponding for the episode and one for the entire podcast. Does that mean that must be 2 dialogs or only one for the episode for the moment? Just to clarify that understood well 😅

ByteHamster commented:
> Ohh you're right! At the beginning, I was a little confused if it should look like the mockup.

I think the main point of #1215 is to make the difference between episode file url and episode website url more clear. Having them in a dialog is easier to see, I think. Just for comparison, I attached a screenshot of the old layout in German. If half of the text is cut off, it is even harder to understand what you actually share. Additionally, the translation is inconsistent. Having it only in one location (like in the mock-up) prevents that.

<img width="250" src="https://user-images.githubusercontent.com/5811634/85606065-12aba100-b653-11ea-8d07-52d1aa12805d.png" />

It does not have to look exactly like the mock-up. Especially the new feature of including the title does not have to be added, for now, but I like the idea of separating "Website/Media file" and "Timestamp/No timestamp".

> Does that mean that must be 2 dialogs or only one for the episode for the moment? Just to clarify that understood well

I think to close #1215, having both would be good. This PR does not have to do both at the same time, though. This can be done in another PR :)

gerardolgvr commented:
Thanks a lot for the information @ByteHamster , I understood your points. It's like you opened my eyes style "Bird box movie"

I wil try first with episode dialog and  see if it goes well according to the objective :)

gerardolgvr commented:
Hello @ByteHamster , I think I almost have the episode dialog but before to finish I would like to ask you about the "title feature", I found some methods at ShareUtils. 

When you mentioned "title". Are you making reference to the podcast title and episode title, aren't you? 

I mean, the objective of the "title" switch is add and remove that titles when are selecting the switch on/off respectively. For example, if I want to share an episode and I'm only select the episode website and I'm not select or turn on the title switch and timestamp switch the result is gonna be only the url, am I right?

ByteHamster commented:
> When you mentioned "title". Are you making reference to the podcast title and episode title, aren't you?

Exactly. I am not sure if it is a good idea to add the title switch, though, because it adds more options. In many cases, having more options is not a good thing.

> https://ben.balter.com/2016/07/21/removing-a-feature-is-a-feature/  
> Every time you provide a user with an option, you’re asking them to make a decision, and those decisions quickly add up, especially when the user doesn’t care or doesn’t fully understand the implications.  [...]
> that new knob you added for User B, because “why not?, it’s just one knob”, is going to make it harder for User A to find the knob they’re actually looking for.

I already like your dialog as it is. Unless you really want to add the title switch, I would just keep it like in your screenshot.

What do you think about the dialog remembering the last choice automatically (so the checkboxes stay checked)? I think this could be a pretty nice feature :)

gerardolgvr commented:
> Exactly. I am not sure if it is a good idea to add the title switch, though, because it adds more options. In many cases, having more options is not a good thing.
 
Yes, could be a little confuse for the final users. I think in that way is easy to understand


> What do you think about the dialog remembering the last choice automatically (so the checkboxes stay checked)? I think this could be a pretty nice feature :)

Sure, I like the idea. Seems well to you save these preferences at Shared Preferences?

gerardolgvr commented:
@ByteHamster  I have added shared preferences to save the last options selected at ShareEpisodeDialog, added some tests.

Feel free to review the changes I made if a change is necessary. 

I would like to help with the other dialog :)

gerardolgvr commented:
![WhatsApp Image 2020-06-27 at 1 00 45 AM](https://user-images.githubusercontent.com/20119298/85915969-b5aa1980-b811-11ea-9e92-d0de53ca492b.jpeg)

At the moment to Share a podcast only exist 2 options. The new dialog should have these but I notice at the mockup are 2 switches:
- author
- title

A this case should contain both (author and title)?

ByteHamster commented on app/src/androidTest/java/de/test/antennapod/ui/DialogsTest.java:
```diff
@@ -0,0 +1,98 @@
+ * User interface tests for queue fragment.
+ */
+@RunWith(AndroidJUnit4.class)
+public class DialogsTest {
```
I would move this to a `dialogs` package and name the class `ShareDialogTest`, so it is more modular

gerardolgvr replied:
yes, you're right. I'm gonna fix it.

ByteHamster commented on app/src/androidTest/java/de/test/antennapod/ui/DialogsTest.java:
```diff
@@ -0,0 +1,98 @@
+        EspressoTestUtils.setLastNavFragment(AddFeedFragment.TAG);
+        activityRule.launchActivity(new Intent());
+
+        String url = "https://omny.fm/shows/silence-is-not-an-option/why-not-being-racist-is-not-enough";
```
We already have a class `UITestUtils` that automatically adds and hosts a podcast. I would prefer that over a real podcast because it makes us less dependent on somebody else. See `PlaybackTest` for example usage.

gerardolgvr replied:
I'm gonna have a look in that to update the tests :)

ByteHamster commented on app/src/androidTest/java/de/test/antennapod/ui/DialogsTest.java:
```diff
@@ -0,0 +1,98 @@
+        onView(allOf(isDisplayed(), withText(R.string.share_label)));
+        Thread.sleep(1000);
+        onView(withText(R.string.share_episode_positive_label_button)).perform(scrollTo()).perform(click());
+        Thread.sleep(2000);
```
Explicit sleep calls are usually not good for tests. It either makes them slow or it makes them fast but fails sometimes. Espresso should theoretically wait for the UI thread automatically.

gerardolgvr replied:
thanks I updated the tests :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,121 @@
+    public ShareDialog(Context ctx, FeedItem item) {
+        this.ctx = ctx;
+        this.item = item;
+        prefs = PreferenceManager.getDefaultSharedPreferences(ctx);
```
Please use a preference store specifically for the dialog. This reduces the probability of name collisions. `ctx.getSharedPreferences (SHARE_DIALOG_PREFS, Context.MODE_PRIVATE)`

gerardolgvr replied:
fixed for the next commit :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,121 @@
+                        if (radioEpisodeWebsite.isChecked()) {
+                            if (switchStartAt.isChecked()) {
+                                ShareUtils.shareFeedItemLink(ctx, item, true);
+                                prefs.edit().putBoolean(UserPreferences.PREF_SHARE_EPISODE_START_AT, true).apply();
```
Please decouple this from the UserPreferences class (that class is already huge). Just create the constants here -- they are not needed globally.

gerardolgvr replied:
perfect, I have already update this for the next commit. Thanks

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,121 @@
+                            } else {
+                                ShareUtils.shareFeedItemDownloadLink(ctx, item);
+                                prefs.edit().putBoolean(UserPreferences.PREF_SHARE_EPISODE_START_AT, false).apply();
+                            }
```
Can you please shorten this? These checks look pretty big and I think they can be simplified, eg. something like this:

```
ShareUtils.shareFeedItemDownloadLink(ctx, item, switchStartAt.isChecked());
prefs.edit().putBoolean(UserPreferences.PREF_SHARE_EPISODE_START_AT, switchStartAt.isChecked()).apply();
```

gerardolgvr replied:
sure, I have already updte this for the next commit. Thanks for this observation

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -765,6 +765,14 @@
+    <string name="share_episode_label">Share…</string>
+    <string name="share_episode_positive_label_button">Share</string>
+    <string name="share_dialog_include_label">Include:</string>
+    <string name="share_start_at_timer_dialog_label">Start at (timer)</string>
```
What about `Playback position`?

gerardolgvr replied:
are you refering to rename "share_start_at_timer_dialog_label"  for "share_playback_position_dialog_label"? 

Sorry , I don't catch this comment. Can you explain a little more please :)

ByteHamster replied:
Sorry, my comments are getting shorter and shorter in the past few weeks. That has lead to some issues with another PR already. I'm pretty busy and try to squeeze in some AntennaPod work.

I meant the text of the string that is shown to the user. I think that "Playback position" might be easier to understand than "Start at (timer)". Not completely sure, though. I think the word "timer" is only used for the sleep timer currently. How does YouTube name that? I think they also use something like "Include position" or "Include timestamp"

gerardolgvr replied:
Thanks, don't worry. 

YouTube just called it "Start at" (without 'timer' word) if you prefer I can delete the word 'timer' or just use "Playback position". I like the last one , I agree, it is pretty easy to understand

ByteHamster replied:
Good point mentioning YouTube. I think we are in a bit different situation here, though. Just including the position with AntennaPod does not automatically mean that the episode is actually played from that position. I have a feeling that it is probably not done in most cases. So "Playback position" could be the better choice, here, I think.

keunes replied:
You mentioned YouTube first, @ByteHamster ;)
I'd put my vote to 'Playback position'

ByteHamster replied:
> You mentioned YouTube first, @ByteHamster ;)

Oops :smile: I currently have a few other things in mind and somehow completely forgot that...

gerardolgvr replied:
no problem. I like "Playback position". I have added it :)

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -765,6 +765,14 @@
 
+    <!-- Share episode dialog -->
+    <string name="share_episode_label">Share…</string>
+    <string name="share_episode_positive_label_button">Share</string>
```
Is there a string that already contains this text? If so, please use the old one instead. This reduces inconsistency when translators translate them differently.

gerardolgvr replied:
Yes, there was a "Share..." string. I updated and I deleted the extra I created before

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/menuhandler/FeedItemMenuHandler.java:
```diff
@@ -243,20 +235,9 @@ public static boolean onMenuItemClicked(@NonNull Fragment fragment, int menuItem
-                ShareUtils.shareFeedItemFile(context, selectedItem.getMedia());
+            case R.id.share_item:
+                new ShareDialog(context, selectedItem)
+                        .show(((AppCompatActivity) context).getSupportFragmentManager(), "ShareEpisodeDialog");
```
Please change the type of `context` instead of casting (`fragment.getActivity()`?). I fear that this could lead to runtime crashes otherwise.

gerardolgvr replied:
thanks, update it. For the next commit

gerardolgvr commented:
> > At the moment to Share a podcast only exist 2 options. The new dialog should have these but I notice at the mockup are 2 switches
> 
> You are right, having a full dialog for 2 switches is probably a bit too much. What about using a sub menu here (like it was in the case of the episodes)? This should be shorter there anyways, so we will not have the problem of translations overflowing the screen.

Sure, I like the idea. I added a submenu. 
Here some screenshots
![WhatsApp Image 2020-06-28 at 5 17 49 PM](https://user-images.githubusercontent.com/20119298/85959849-85bf5b00-b964-11ea-8457-f7d88e27520d.jpeg)
![WhatsApp Image 2020-06-28 at 5 17 49 PM (1)](https://user-images.githubusercontent.com/20119298/85959851-88ba4b80-b964-11ea-807e-46644d653e78.jpeg)

If there's any change or suggestion let me know :)

ByteHamster commented:
> Sure, I like the idea. I added a submenu.

I think the word "Share" in front of the menu items is no longer needed, now that they are in a submenu. Removing the word also makes the text shorter. We could maybe think about a different wording for "Website URL" vs "Podcast URL". I think that this is pretty hard to understand. What do you think about "Podcast feed link" and "Website link"? I think many users do not know the word "URL".

keunes commented:
> We could maybe think about a different wording for "Website URL" vs "Podcast URL". I think that this is pretty hard to understand. What do you think about "Podcast feed link" and "Website link"? I think many users do not know the word "URL".

"Website URL" vs "Podcast URL" are confusing indeed. I think 'URL' is pretty known (at least in English). And I don't like 'link' (also, it's not a link, semantically). What about "Website" and "Podcast feed", or "Website address" and "Podcast feed URL"?

ByteHamster commented:
> "Website address" and "Podcast feed URL"?

I like the word "address" :+1: It's correct on a technical level and easier to understand than "URL".

> And the strings for the episode sharing should be "Episode webpage" (rather than 'Episode website').

:+1: I did not even know that there is a difference. In German, both words are the same.

gerardolgvr commented:
I updated the strings , some screenshots:

![WhatsApp Image 2020-06-30 at 11 14 59 PM](https://user-images.githubusercontent.com/20119298/86202428-cd421480-bb27-11ea-979d-ab2adcfae0c1.jpeg)

![WhatsApp Image 2020-06-30 at 11 14 59 PM (1)](https://user-images.githubusercontent.com/20119298/86202435-d3d08c00-bb27-11ea-9435-68f73c7f98af.jpeg)

ByteHamster commented on app/src/androidTest/java/de/test/antennapod/dialogs/ShareDialogTest.java:
```diff
@@ -0,0 +1,94 @@
+import static org.hamcrest.CoreMatchers.allOf;
+
+/**
+ * User interface tests for queue fragment.
```
The comment is wrong :)

gerardolgvr replied:
fixed :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,117 @@
+    private Switch switchStartAt;
+    private SharedPreferences prefs;
+
+    public ShareDialog(Context ctx, FeedItem item) {
```
Instead of passing parameters, please use arguments (see https://stackoverflow.com/a/15459259/4193263). This prevents issues when the system recreates the fragment (after the app was in background, for example). We use the `newInstance` pattern in other parts of the app, too, so this would also be consistent.

gerardolgvr replied:
thanks, updated it. 

I didn't notice this. I was using SleepTimerDialog as example. I have updated it

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,117 @@
+    public ShareDialog(Context ctx, FeedItem item) {
+        this.ctx = ctx;
+        this.item = item;
+        prefs = ctx.getSharedPreferences("SHARE_DIALOG_PREFS", Context.MODE_PRIVATE);
```
Please use `ShareDialog` for the name. This is more consistent with the other preference files.

gerardolgvr replied:
added :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,117 @@
+        setupOptions();
+
+        builder
+                .setPositiveButton(R.string.share_label, new DialogInterface.OnClickListener() {
```
Could you please remove the newline between `builder` and `setPositiveButton` and move the indentation of the listener to the left? I think checkstyle should be happy then, too, and it uses less horizontal space.

gerardolgvr replied:
yup, at the beginning I moved to a new line because I was worried about the length but I checked at the length of the line is 96, then I updated it .

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,117 @@
+                            prefs.edit().putBoolean(PREF_SHARE_EPISODE_START_AT, switchStartAt.isChecked()).apply();
+
+                            prefs.edit().putBoolean(PREF_SHARE_EPISODE_WEBSITE, false).apply();
+                            prefs.edit().putBoolean(PREF_SHARE_EPISODE_MEDIA, true).apply();
```
Are those two preferences exclusive? If yes, it could be better to only store one preference.

gerardolgvr replied:
I decided to use a String instead of use 2 booleans, in this way I only have to check the value of the string and then activate or desactived the radio buttons :) . The code now is more clear and easy to understand. I'm gonna add the new commit

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -136,12 +136,12 @@
     <string name="rename_feed_label">Rename podcast</string>
     <string name="remove_feed_label">Remove podcast</string>
-    <string name="share_label">Share&#8230;</string>
+    <string name="share_label">Share</string>
```
I think it would be better to have 2 strings here. One with the dots and one without. Menu items with sub menus usually have dots in their title but the dialog title and button are better without dots. Let's wait for a reply by @keunes first. @keunes, what do you think?

keunes replied:
Yup, agreed.

gerardolgvr replied:
Yup, I like the idea too

ByteHamster commented on app/src/main/res/layout/share_episode_dialog.xml:
```diff
@@ -0,0 +1,40 @@
+        android:orientation="vertical">
+
+        <RadioButton android:id="@+id/share_episode_website_radio"
+            android:layout_width="wrap_content"
```
If you set the width to `match_parent`, the buttons get easier to tap (while looking exactly the same).

gerardolgvr replied:
Added :)

ByteHamster commented on app/src/main/res/layout/share_episode_dialog.xml:
```diff
@@ -0,0 +1,40 @@
+        android:id="@+id/share_start_at_timer_dialog"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/share_playback_position_dialog_label" />
```
I think the dialog could look more balanced if this was a CheckBox instead of a Switch. Then all the blue elements line up on the left. I am not completely sure, though. What do you think?

gerardolgvr replied:
Yes I agree, I will try to update it.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,117 @@
+                    @Override
+                    public void onClick(DialogInterface dialog, int id) {
+                        if (radioEpisodeWebsite.isChecked()) {
+                            ShareUtils.shareFeedItemLink(ctx, item, switchStartAt.isChecked());
```
I think this shares the same link regardless of the checkbox now. Copy-paste error?

`switchStartAt.isChecked()` is called quite a few times. What about adding a variable `boolean includePosition = switchStartAt.isChecked()`?

gerardolgvr replied:
Sure! You're right. I have added it to the new commit. :)

gerardolgvr commented:
I hope to review and solve your comments today :)

ByteHamster commented on app/src/main/res/menu/mediaplayer.xml:
```diff
@@ -63,27 +63,5 @@
         android:menuCategory="container"
         custom:showAsAction="never"
         android:title="@string/share_label">
```
I would use `share_label_with_ellipses` here

gerardolgvr replied:
yup, adde it

ByteHamster commented on app/src/main/res/menu/feeditem_options.xml:
```diff
@@ -72,28 +72,6 @@
         android:id="@+id/share_item"
         android:menuCategory="container"
         android:title="@string/share_label">
```
I would use `share_label_with_ellipses` here

gerardolgvr replied:
yup, I forgot update this . Thanks

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/ShareDialog.java:
```diff
@@ -0,0 +1,121 @@
+                prefs.edit().putBoolean(PREF_SHARE_EPISODE_START_AT, includePlaybackPosition).apply();
+            }
+        })
+        .setNegativeButton(R.string.cancel_label, new DialogInterface.OnClickListener() {
```
Checkstyle complains here. If you just remove the line break between `)` and `.`, it is happy :)

gerardolgvr replied:
fixed :)

gerardolgvr commented:
I have the changes and made a new commit :)

ByteHamster commented:
Awesome, thanks! The change will be available in version 2.1.0 (not in 2.0.0 because that version is already in the beta stage).

gerardolgvr commented:
Thank you! :)
